lib/vsc/mympirun/mpi/mpi.py

```diff
@@ -198,6 +198,42 @@ def _loop_process_output(self, output):
             self.log.error(warning_msg + '\n')
             self.seen_output = True
 
+class RunAsyncMPI(RunAsyncLoopStdout):
+    """
+    RunAsync for MPI in case a program 'hangs'
+
+    """
+    def __init__(self, cmd, **kwargs):
+        self.output_timeout = kwargs.pop('output_timeout', None)
+        # check timeout
+        try:
+            self.output_timeout = int(self.output_timeout)
+        except ValueError as err:
+            raise ValueError("Invalid timeout value: %s (should be integer): %s" % (self.output_timeout, err))
+
+        super(RunAsyncMPI, self).__init__(cmd, **kwargs)
+
+        self.seen_output = self.output_timeout < 0 #no check when output_timeout is negative
+
+    def _loop_process_output(self, output):
+        """ Send output to stdout + hang check """
+        if len(output) > 0:
+            self.seen_output = True
+
+        time_passed = self._loop_count * self.LOOP_TIMEOUT_MAIN
+        if time_passed > self.output_timeout and not self.seen_output:
+            warning_msg = '\n'.join([
+            "WARNING: mympirun has been running for %s seconds without seeing any output." % time_passed,
+            "This may mean that your program is hanging, please check and make sure that is not the case!",
+            '',
+            "If this warning is printed too soon and the program is doing useful work without producing any output,",
+            "you can increase the timeout threshold via --output-check-timeout (current setting: %s seconds)" % self.output_timeout,
+            ])
+            self.log.error(warning_msg + '\n')
+            self.seen_output = True
```

Comment:
same here, don't copy-paste this, hoist it into a function you call to that produces the warning message (but don't pass down `self.log`, do that locally)